### PR TITLE
proposal(contrib/openstack): terraform replacement for shell script

### PR DESCRIPTION
**Pull request by paulczar**
The existing provisioning script for openstack is fairly brittle and relies
on a bunch of software that the user needs to install ( python-novaclient etc )

This is a working [just] prototype that uses the terraform.

```

### CODE REVIEW

**Comment by deis-admin** [comment]
Your Pull Request can earn you credit at DigitalOcean! See http://docs.deis.io/en/latest/contributing/hacking/#digitalocean-credit for details.

Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

**Comment by paulczar** [comment]
this is not meant for immediate merging ...   but to open a discussion and see if this is wanted, and/or what other features would be interesting.    Documentation would also need to be updated in the correct place before/after merge.

**Comment by sgoings** [comment]
I really appreciate the effort put into this PR - I think this work could integrate very nicely into [rigger](https://github.com/deis/rigger) which is intended to provide a common user interface (currently only CLI) on top of our cloud provider provisioning scripts. Take a look at how I've defined the interaction between rigger and the deis provisioning scripts in [this design document](https://github.com/deis/deis/issues/4345).

At the current time we don't test the OpenStack-oriented Deis provisioning process, but it would be most excellent if we started doing that, and this PR could jumpstart that effort on our end. :+1:

**Comment by paulczar** [comment]
Yeah we can probably use terraform if its openstack support is up to scratch.   do we have an example where we're using it for another platform that I can borrow style etc from ?

**Comment by bacongobbler** [comment]
@paulczar #4315 :)

**Comment by kalbasit** [comment]
Accidentally closed the PR.

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

Your Pull Request can earn you credit at DigitalOcean! See http://docs.deis.io/en/latest/contributing/hacking/#digitalocean-credit for details.

**Comment by deis-admin** [comment]
Thanks for the contribution! Please ensure your commits follow our [style guide](http://docs.deis.io/en/latest/contributing/standards/#commit-style-guide). This code will be tested once a Deis maintainer reviews it.

Your Pull Request can earn you credit at DigitalOcean! See http://docs.deis.io/en/latest/contributing/hacking/#digitalocean-credit for details.

**Comment by carmstrong** [comment]
@paulczar Did you get a chance to take a look at Terraform support for OpenStack? Docs are here: https://terraform.io/docs/providers/openstack/index.html

I'm not familiar enough with OpenStack to judge the comprehensiveness of support, but I agree with @bacongobbler that if we're going to rewrite a provider, we should probably do it in Terraform whenever possible.

**Comment by paulczar** [comment]
I hope to find the time this week to take a shot at doing it with terraform.

**Comment by paulczar** [comment]
first pass at terrorform support.   seems pretty straightforward.   not fully tested yet ... but I do end up with working coreos instances...    updated the PR to provide status.

@sgoings can you glance over the rigger bits?  I just blank copy and pasted from the DO section and made a best guess.

**Comment by sgoings** [comment]
Sweet! Thanks @paulczar for putting time into this! I'll take a closer look tomorrow and see what I can do to add OpenStack as a CI-tested provider soon...

File: contrib/openstack/README.md [hunk 1]
```diff
@@ -1,3 +1,14 @@
...
+
+```
+$ make discovery-url
+$ terraform apply \
+      -var "username=$OS_USERNAME" \
+      -var "password=$OS_PASSWORD" \
+      -var "tenant=$OS_TENANT_NAME" \
+      -var "auth_url=$OS_AUTH_URL" \
+      -var "image=coreos-717.3.0" \
+      contrib/openstack
```
Comment (sgoings) [inline contrib/openstack/README.md]: I think you might be missing a few vars in here... this README.md should also ideally point to the official docs (installing Deis on OpenStack) once this PR gets finalized.

File: contrib/openstack/config.sh [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+prompt "What glance image should I use? [coreos-stable]" GLANCE_IMAGE
+prompt "What flavor ? [2]" FLAVOR
+prompt "Path to your public key: [~/.ssh/id_rsa.pub] " KEY
+prompt "What is the name of your neutron network? [internal]" NETWORK
+prompt "What is the name of your floating IP pool ? [external]" POOL
+prompt "what is your openstack auth URL? [${OS_AUTH_URL}]" OS_AUTH_URL
+prompt "what is your openstack username? [${OS_USERNAME}]" OS_USERNAME
+prompt "what is your openstack password? [**********]" OS_PASSWORD
+prompt "what is your openstack tenant name? [${OS_TENANT_NAME}]" OS_TENANT_NAME
```
Comment (sgoings) [inline contrib/openstack/config.sh]: Prompt takes 3 variables:

```
prompt "<question>" <return_var> <default suggested var>
```

or just look at the [implementation here](https://github.com/deis/rigger/blob/master/lib/interactive.sh#L52).

File: contrib/openstack/config.sh [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+export TF_VAR_image="${GLANCE_IMAGE}"
+export TF_VAR_flavor="${FLAVOR}"
+export TF_VAR_public_key_path="${KEY}"
+export TF_VAR_network_name="${NETWORK}"
+export TF_VAR_floatingip_pool="${POOL}"
+export TF_VAR_auth_url="${OS_AUTH_URL}"
+export TF_VAR_username="${OS_USERNAME}"
+export TF_VAR_password="${OS_PASSWORD}"
+export TF_VAR_tenant_name="${OS_TENANT_NAME}"
+export TF_VAR_stack_name="deis-${DEIS_ID}"
```
Comment (sgoings) [inline contrib/openstack/config.sh]: You should be able to set most of these directly via the prompt function above. Then you just have to export them here.

File: contrib/openstack/config.sh [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+                 TF_VAR_image \
+                 TF_VAR_flavor \
+                 TF_VAR_public_key_path \
+                 TF_VAR_network_name \
+                 TF_VAR_floatingip_pool \
+                 TF_VAR_auth_url \
+                 TF_VAR_username \
+                 TF_VAR_password \
+                 TF_VAR_tenant_name \
+                 TF_VAR_stack_name
```
Comment (sgoings) [inline contrib/openstack/config.sh]: I think you should only save the TF_VAR_* types of the variables... especially if you make the change suggested above where prompt is setting the TF_VAR_* for you.

**Comment by carmstrong** [comment]
ping @sgoings for re-review -- looks like @paulczar made your last requested change

**Comment by mboersma** [comment]
Jenkins, add to whitelist.

File: contrib/openstack/README.md [hunk 5]
```diff
@@ -1,3 +1,16 @@
-# Provision a Deis Cluster on OpenStack
+# Provision a Deis Cluster on Openstack
```
Comment (mboersma) [inline contrib/openstack/README.md]: This should stay as `OpenStack` with InterCaps.

File: contrib/openstack/README.md [hunk 6]
```diff
@@ -1,3 +1,16 @@
...
 
 Please refer to the instructions at http://docs.deis.io/en/latest/installing_deis/openstack/.
+
+Meanwhile here is how I've been testing it:
```
Comment (mboersma) [inline contrib/openstack/README.md]: This sounds too informal even for our contrib/ docs--could you rephrase it as something like `Provision a cluster with a command such as:`

File: contrib/openstack/README.md [hunk 7]
```diff
@@ -1,3 +1,16 @@
...
+
+```
+$ make discovery-url
+$ terraform apply \
+      -var "username=$OS_USERNAME" \
+      -var "password=$OS_PASSWORD" \
+      -var "tenant=$OS_TENANT_NAME" \
+      -var "auth_url=$OS_AUTH_URL" \
+      -var "image=coreos-717.3.0" \
```
Comment (mboersma) [inline contrib/openstack/README.md]: Could you update this to 766.4.0 (or whatever is current in https://github.com/deis/deis/blob/master/contrib/utils.sh#L16)?

**Comment by paulczar** [comment]
rebased it to remove all the clutter from the original python stuff.

File: contrib/openstack/variables.tf [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+}
+
+variable "public_key_path" {
+  description = "The path of the ssh pub key"
+  default = "~/.ssh/id_rsa.pub"
+}
+
+variable "image" {
+  description = "the image to use"
+  default = "coreos-717.3.0"
```
Comment (carmstrong) [inline contrib/openstack/variables.tf]: Should we bump this to 835.11.0?

Comment (mboersma) [inline contrib/openstack/variables.tf]: Or 835.13.0?

**Comment by mboersma** [comment]
@sgoings can you give this a look-see when time permits? I'm trying to fish or cut bait on remaining LTS release issues soon. :clock1130:

**Comment by mboersma** [comment]
OpenStack testing efforts are now targeted at Deis v2. It's unlikely we'll revive this PR so I'm closing it. Thanks again @paulczar!
